File: ios-app/TomsDiary/Services/HandwritingRenderer.swift
//
//  HandwritingRenderer.swift
//  Tom's Diary - iOS
//
//  Renders text as handwritten images using the Caveat font.
//  Supports word-by-word streaming for smooth display.
//

import Foundation
import UIKit

final class HandwritingRenderer {
    
    // MARK: - Properties
    
    private var caveatFont: UIFont?
    
    // MARK: - Initialization
    
    init() {
        loadFont()
    }
    
    private func loadFont() {
        // Try to load from bundle (registered in Info.plist)
        caveatFont = UIFont(name: "Caveat-Regular", size: 90)
        
        if caveatFont == nil {
            // Try to load from assets
            if let path = Bundle.main.path(forResource: "Caveat-Regular", ofType: "ttf"),
               let data = try? Data(contentsOf: URL(fileURLWithPath: path)),
               let font = UIFont(data: data) {
                caveatFont = font
            }
        }
        
        if caveatFont == nil {
            // Last resort: use system script font
            caveatFont = .systemFont(ofSize: 90, weight: .regular)
            print("⚠️ Warning: Caveat font not found, using system font")
        }
    }
    
    // MARK: - Configuration
    
    struct RenderOptions {
        let fontSize: CGFloat
        let backgroundColor: UIColor
        let textColor: UIColor
        let padding: CGFloat
        let maxWidth: CGFloat
        let addVariation: Bool
        
        static let `default` = RenderOptions(
            fontSize: 90,
            backgroundColor: .white,
            textColor: .black,
            padding: 20,
            maxWidth: 900,
            addVariation: true
        )
    }
    
    // MARK: - Public Methods
    
    /// Calculates appropriate font size for given screen width
    func calculateFontSize(for screenWidth: CGFloat) -> CGFloat {
        max(48, min(120, screenWidth / 14))
    }
    
    /// Renders a single word as an image
    func renderWord(_ word: String, options: RenderOptions = .default) -> UIImage? {
        guard let font = caveatFont else { return nil }
        
        let attributes: [NSAttributedString.Key: Any] = [
            .font: font,
            .foregroundColor: options.textColor
        ]
        
        let wordSize = word.size(withAttributes: attributes)
        
        // Get font metrics
        let fontDesc = font.fontDescriptor
        let ctFont = CTFontCreateWithFontDescriptor(fontDesc, options.fontSize, nil)
        let descender = CTFontGetDescender(ctFont)
        let ascender = CTFontGetAscender(ctFont)
        let wordHeight = abs(descender - ascender)
        
        let leftPadding: CGFloat = 15
        let rightPadding: CGFloat = 30
        let topPadding: CGFloat = 10
        let bottomPadding: CGFloat = 25
        
        let canvasWidth = wordSize.width + leftPadding + rightPadding
        let canvasHeight = wordHeight + topPadding + bottomPadding
        
        let rect = CGRect(x: 0, y: 0, width: canvasWidth, height: canvasHeight)
        
        UIGraphicsBeginImageContextWithOptions(rect.size, false, 0)
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        
        // Draw background
        options.backgroundColor.setFill()
        context.fill(rect)
        
        // Calculate text position (baseline)
        let x = leftPadding
        let y = topPadding + wordHeight - descender
        
        // Add variation if enabled
        if options.addVariation {
            // Small rotation in radians (not degrees)
            let rotation = Double.random(in: -0.5...0.5) * 0.02  // ±0.01 radians (~±0.57 degrees)
            let xOffset = Double.random(in: -0.5...0.5) * 2  // ±1 point offset
            
            context.save()
            context.translateBy(x: x + CGFloat(xOffset), y: y)
            context.rotate(by: CGFloat(rotation))
            
            let attributedString = NSAttributedString(string: word, attributes: attributes)
            attributedString.draw(at: .zero)
            
            context.restore()
        } else {
            let attributedString = NSAttributedString(string: word, attributes: attributes)
            attributedString.draw(at: CGPoint(x: x, y: y))
        }
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return image
    }
    
    /// Renders words from text as a sequence of images
    func renderWordStream(_ text: String, options: RenderOptions = .default) -> AsyncStream<UIImage> {
        AsyncStream { continuation in
            let words = text.split(separator: " ").map(String.init)
            
            for word in words {
                if let image = renderWord(word, options: options) {
                    continuation.yield(image)
                }
            }
            
            continuation.finish()
        }
    }
}

File: android-app/app/src/main/java/com/tomsdiary/HandwritingRenderer.kt
package com.tomsdiary

import android.content.Context
import android.graphics.*
import java.io.ByteArrayOutputStream
import kotlin.random.Random

class HandwritingRenderer(private val context: Context) {
    
    private var caveatFont: Typeface? = null
    
    init {
        loadFont()
    }
    
    private fun loadFont() {
        try {
            caveatFont = Typeface.createFromAsset(context.assets, "Caveat-Regular.ttf")
            android.util.Log.d("HandwritingRenderer", "Caveat font loaded successfully")
        } catch (e: Exception) {
            android.util.Log.e("HandwritingRenderer", "Failed to load Caveat font", e)
            caveatFont = Typeface.create("Caveat", Typeface.NORMAL)
        }
    }
    
    data class RenderOptions(
        val fontSize: Int = 90,
        val backgroundColor: Int = Color.WHITE,
        val textColor: Int = Color.BLACK,
        val padding: Int = 20,
        val maxWidth: Int = 900,
        val addVariation: Boolean = true
    )
    
    data class WordRenderResult(
        val bitmap: Bitmap,
        val width: Int,
        val height: Int,
        val x: Float,
        val y: Float
    )
    
    fun calculateFontSize(screenWidth: Int): Int {
        return kotlin.math.max(48, kotlin.math.min(120, screenWidth / 14))
    }
    
    fun renderWord(word: String, options: RenderOptions = RenderOptions()): Bitmap {
        val font = caveatFont ?: Typeface.create("Caveat", Typeface.NORMAL)
        val paint = Paint().apply {
            typeface = font
            textSize = options.fontSize.toFloat()
            color = options.textColor
            isAntiAlias = true
            style = Paint.Style.FILL
        }
        
        val fontMetrics = paint.fontMetrics
        
        val wordWidth = paint.measureText(word)
        val wordHeight = kotlin.math.abs(fontMetrics.bottom - fontMetrics.top)
        
        val leftPadding = 15f
        val rightPadding = 30f
        val topPadding = 10f
        val bottomPadding = 25f
        
        val canvasWidth = (wordWidth + leftPadding + rightPadding).toInt()
        val canvasHeight = (wordHeight + topPadding + bottomPadding).toInt()
        
        val bitmap = Bitmap.createBitmap(canvasWidth, canvasHeight, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        
        canvas.drawColor(options.backgroundColor)
        
        val x = leftPadding
        val y = topPadding + wordHeight - fontMetrics.bottom
        
        if (options.addVariation) {
            val rotation = (Random.nextDouble() - 0.5).toFloat() * 0.02f
            val xOffset = (Random.nextDouble() - 0.5).toFloat() * 2f
            
            canvas.save()
            canvas.translate(x + xOffset, y)
            canvas.rotate((rotation * 180f / kotlin.math.PI).toFloat())
            canvas.drawText(word, 0f, 0f, paint)
            canvas.restore()
        } else {
            canvas.drawText(word, x, y, paint)
        }
        
        return bitmap
    }
    
    fun renderWordStream(
        text: String,
        screenWidth: Int,
        options: RenderOptions = RenderOptions(fontSize = calculateFontSize(screenWidth))
    ): Sequence<WordRenderResult> {
        val words = text.split(' ')
        var currentX = options.padding.toFloat()
        var currentY = (options.padding + options.fontSize).toFloat()
        val lineHeight = (options.fontSize * 1.5f)
        
        return sequence {
            for (word in words) {
                val bitmap = renderWord(word, options)
                
                yield(
                    WordRenderResult(
                        bitmap = bitmap,
                        width = bitmap.width,
                        height = bitmap.height,
                        x = currentX,
                        y = currentY
                    )
                )
                
                val wordWidth = paintWordWidth(word, options.fontSize)
                currentX += wordWidth + 10f
                
                if (currentX + wordWidth > options.maxWidth - options.padding) {
                    currentX = options.padding.toFloat()
                    currentY += lineHeight
                }
            }
        }
    }
    
    private fun paintWordWidth(word: String, fontSize: Int): Float {
        val font = caveatFont ?: Typeface.create("Caveat", Typeface.NORMAL)
        val paint = Paint().apply {
            typeface = font
            textSize = fontSize.toFloat()
        }
        return paint.measureText(word)
    }
    
    fun bitmapToBase64(bitmap: Bitmap): String {
        val byteArrayOutputStream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream)
        val byteArray = byteArrayOutputStream.toByteArray()
        return byteArray.encodeBase64().decodeToString()
    }
    
    companion object {
        private fun ByteArray.encodeBase64(): ByteArray {
            return android.util.Base64.encode(this, android.util.Base64.NO_WRAP)
        }
    }
}
